probe-rs/src/debug/registers.rs

```diff
@@ -1,162 +1,243 @@
-use crate::core::Core;
-
-use probe_rs_target::Architecture;
-
-use std::collections::HashMap;
-
-use crate::core::{RegisterDescription, RegisterFile};
+use crate::{
+    core::{Core, RegisterDataType, RegisterFile},
+    Error, RegisterId, RegisterValue,
+};
+
+/// The group name of a register.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum RegisterGroup {
+    /// Core / CPU Registers
+    Base,
+    /// Argument Register
+    Argument,
+    /// Result Register
+    Result,
+    /// [`RegisterFile`] contains some register descriptions that are not part of an array, and may or may not have the same `RegisterId` as registers in other groups.
+    Singleton,
+}
 
-/// All the register information currently available.
+/// Stores the relevant information from [`RegisterDescription`] as well as additional information required during debug.
 #[derive(Debug, Clone, PartialEq)]
-pub struct Registers {
-    pub(crate) register_description: &'static RegisterFile,
+pub struct DebugRegister {
+    /// To lookup platform specific details of register definitions.
+    pub register_file: &'static RegisterFile,
+    /// Register definitions are grouped depending on their purpose.
+    pub group: RegisterGroup,
+    /// The name of the register.
```

Comment:
Why not capture a reference to the `RegisterDescription` so you don't have to duplicate data / logic?  For example, `size_in_bytes` could delegate to something like `self.reg_desc.size_in_bytes()`

Reply:
Good catch. I've added a TODO for my next round.